introduce cache bypass to fix issues with invalid store configs

mwr opened the pull request:
This pull-request fixes issues with the Service TransactionStatus Process which run using a Cronjob.

From time to time it can happen, that you get an error message much like this one.
"Payment method configuration with id "18" not found."

This error is thrown in ``Observer_TransactionStatus_InvoiceCreate`` but the root cause is to be found during config initialisation. 
During processing the pending TransactionStatus, the config is first loaded here: \Payone_Core_Model_Service_TransactionStatus_Process::execute:86

In some circumstances the wrong configuration is returned. For example the order is for store-id 2 but the config that is returned is not. The root of the problem seems to be the caching though this pull request will not address any of the caching logic. 
There are no tests and the Configuration is loaded and used in Frontend/ Admin and Cronjob which could lead to a whole lot of side effects when changing something here.

If there are other occasions where the wrong configuration is loaded let me know, maybe it might make sense to change the caching in some way.

So what this pull-request actually does is checking if the config.store-id matches the order.store-id and if not bypassing the config-cache and loading the config for the required store-id fresh and without caching it.

Therefore it introduces an additional parameter ``$useCache`` to the Config Service and the ConfigHelper.
It also uses that in Payone_Core_Model_Service_TransactionStatus_Process::execute when the store-ids mismatch.

Furthermore there is also additional logging in ``Observer_TransactionStatus_InvoiceCreate `` in case the error occurs anyways. The InvoiceCreate Observer also has a fallback implemented. (actually I implemented that one first, but thought it might still come in handy.

We are running this patch in production for a multistore setup and so far havn't had any issues.

To test this you would also need a multistore setup, with different payment methods configured for each store.

To check the logs I used ``cat var/log/payone_cron.log | grep "Process - Config loaded with config.store-id=,"``. This command greps the log for the output introduced in this pull-request. Make sure Payone Cron log is activated in config.

mwr commented:
One more thing, I have left some @todos in code for future reference should the errors keep occuring.

fjbender commented:
I have waited a long time to type these letters: Fixes #8 

Thanks! However, as can be seen from the discussion in #8, this is difficult to test. We'll see if we can get someone to try this.

mwr commented:
Just one thing came to my mind: 
The solution I implemented in this Pull-Request could probably be implemented inside the ``\Payone_Core_Model_Service_InitializeConfig::execute`` where the actual loading is done.

Whenever a config is loaded from registry or cache we could double check if the config->getStoreId() matches the one provided to the method in $storeId.

```php

            $config = $helperRegistry->registry($registryKey);
            if ($config instanceof Payone_Core_Model_Config_Interface) {
                // @todo check if store-id from config matches $storeId, if not, discard cache
                return $config;
            }

            $config = $this->loadFromCache();
            if ($config instanceof Payone_Core_Model_Config_Interface) {
                // @todo check if store-id from config matches $storeId, if not, discard cache
                $helperRegistry->register($registryKey, $config);
                return $config;
            }
```

BoulangerV commented on app/code/community/Payone/Core/Model/Service/InitializeConfig.php:
```diff
@@ -55,21 +55,24 @@ class Payone_Core_Model_Service_InitializeConfig
     {
         $this->setStoreId($storeId);
 
         $helperRegistry = $this->helperRegistry();
```
As this line and the following one are not used anymore out of the if-statement, maybe we could move them inside the statement as well ?

mwr replied:
Good point. I moved helperRegistry and registryKey inside the if statement

BoulangerV commented on app/code/community/Payone/Core/Model/Service/InitializeConfig.php:
```diff
@@ -55,21 +55,24 @@ class Payone_Core_Model_Service_InitializeConfig
         $this->setStoreId($storeId);
 
         $helperRegistry = $this->helperRegistry();
         $registryKey = $this->getConfigRegistryKey($storeId);
```
As this line and the previous one are not used anymore out of the if-statement, maybe we could move them inside the statement as well ?

mwr replied:
Good point. I moved helperRegistry and registryKey inside the if statement

mwr commented:
Before we finish this pull-request, I have one more suggestion. 
It's about what I wrote in an earlier comment:

> Just one thing came to my mind:
> The solution I implemented in this Pull-Request could probably be implemented inside the `\Payone_Core_Model_Service_InitializeConfig::execute` where the actual loading is done.
> 
> Whenever a config is loaded from registry or cache we could double check if the config->getStoreId() matches the one provided to the method in $storeId.
> 
> ```
> 
>             $config = $helperRegistry->registry($registryKey);
>             if ($config instanceof Payone_Core_Model_Config_Interface) {
>                 // @todo check if store-id from config matches $storeId, if not, discard cache
>                 return $config;
>             }
> 
>             $config = $this->loadFromCache();
>             if ($config instanceof Payone_Core_Model_Config_Interface) {
>                 // @todo check if store-id from config matches $storeId, if not, discard cache
>                 $helperRegistry->register($registryKey, $config);
>                 return $config;
>             }
> ```

What if we implemented the ``\Payone_Core_Model_Service_InitializeConfig::execute`` method like this:

```php

    /**
     * Generates the Configuration Object for Payone Settings
     *
     * Each StoreId will have an Object respresenting its Configuration
     * The Config Object will be cached
     *
     * @param int $storeId
     * @return Payone_Core_Model_Config_Interface
     */
    public function execute($storeId = null, $useCache = true)
    {
        $this->setStoreId($storeId);

        if($useCache === true) {
            $helperRegistry = $this->helperRegistry();
            $registryKey = $this->getConfigRegistryKey($storeId);

            $config = $helperRegistry->registry($registryKey);
            if (($config instanceof Payone_Core_Model_Config_Interface) && $config->getStoreId() === $storeId) {
                return $config;
            }

            $config = $this->loadFromCache();
            if (($config instanceof Payone_Core_Model_Config_Interface) && $config->getStoreId() === $storeId) {
                $helperRegistry->register($registryKey, $config);
                return $config;
            }
        }

        // … skipped this for a better overview

        // Caching
        if ($useCache === true) {
            $this->saveToCache($config);
        }

        return $config;
    }
```

This way the cache would only be used if the storeIds match. Basically the same thing this pull-request implements in ``\Payone_Core_Model_Service_TransactionStatus_Process::execute``:

```php 
        // if store-ids differ, reload config bypassing the cache
        $this->helper()->logCronjobMessage("ID: {$transactionStatus->getId()} - Process - Config loaded with config.store-id={$config->getStoreId()}, order.store-id: {$order->getStoreId()}", $order->getStoreId());
        if ($config->getStoreId() != $order->getStoreId()) {
            $config = $this->helperConfig()->getConfigStore($order->getStoreId(), false);
            // @todo we should stop processing here, this will only produce errors when working with the wrong config
        }
```

The consequence would be a fall-through the if and loading the config and storing it in the cache, this time with the correct values, since the store-ids are compared and match.

Since we have not located the exact point where the config is stored / loaded faulty, this might be a good way to at least make sure the loaded config store-id matches the store-id passed to the method.
This would make some of the code checking the store-id outside of the InitConfig Service superfluous, but I does not hurt to have double checks I guess.
On the other hand we might solve some issues with other tickets, as @fjbender mentioned this PR in other tickets. #8 

What do you guys think?

The worst thing that could happen is that we load the config too much in some occasions, but I guess it is better than to have issues afterwards.

fjbender commented:
Sounds sensible to me. If @BoulangerV approves, we should go down this road.

BoulangerV commented:
Hi Matthias and Florian,

Sounds good to me as well. This way we should have fresh and correct config when we need it.

I would just advise to check the performances during the test phase, as we are dealing with Magento (:bomb: :smile:), and repeated config loading/cache-bypassing might have an impact, at least on (s)lower hardware environnments.

Otherwise, I'm totally ok with this solution.

Best,
Vincent

fjbender commented:
Thanks for the quick feedback!

@mwr: Will you add this to the PR, or should we try?

mwr commented:
I'll add it, I more ore less already prepared the code :D

mwr commented:
I just pushed the changes which should make sure the config loaded is matching the store-id to the one passed as a parameter.

> I would just advise to check the performances during the test phase, as we are dealing with Magento (💣 😄), and repeated config loading/cache-bypassing might have an impact, at least on (s)lower hardware environnments.

True. It might have a minor impact on performance, but since it is only done when the store-ids mismatch, the impact should not be to hard. Also this case does not occur that much when compared to the successful loads. At least what I was able to extract from the logs from my clients system.

I will try to push those changes to production this week to see the result.

fjbender commented:
@mrw were you able to push the changes to production? Can you share any results? Thanks! :)

mrw commented:
@fjbender I think that was meant for @mwr 😄 - hopefully this message gets to him!

mwr commented:
It got to me thanks ;-) 
Well I haven't found the time to install the patched module to the system yet, was busy with Magento Upgrades and more pressing matters. 
To test this, we would have to upgrade the whole PAYONE module, since we are using an older version. 
I am thinking about extracting the changed files as patch and applying that.
I'll let you know when we tested the changes

fjbender commented:
We were now able to put a quality engineer towards it, and it did not end that well. I'm not too confident with the inner works of the Magento registry, but seems to me we take a wrong turn in `\Payone_Core_Model_Service_InitializeConfig::execute` and end up with an exception:

```
a:5:{i:0;s:55:"Mage registry key "payone_core_config_1" already exists";i:1;s:3313:"#0 /var/www/shops/magento/1942ds/app/Mage.php(229): Mage::throwException('Mage registry k...')
#1 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/Helper/Registry.php(62): Mage::register('payone_core_con...', Object(Payone_Core_Model_Config), false)
#2 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/Model/Service/InitializeConfig.php(75): Payone_Core_Helper_Registry->register('payone_core_con...', Object(Payone_Core_Model_Config))
#3 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/Helper/Config.php(46): Payone_Core_Model_Service_InitializeConfig->execute(NULL, true)
#4 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/Helper/Config.php(83): Payone_Core_Helper_Config->getConfigStore(NULL)
#5 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/Model/Payment/Method/Abstract.php(364): Payone_Core_Helper_Config->getConfigPayment(NULL)
#6 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/Model/Payment/Method/Abstract.php(116): Payone_Core_Model_Payment_Method_Abstract->getConfigPayment()
#7 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/Helper/Payolution.php(145): Payone_Core_Model_Payment_Method_Abstract->isAvailable()
#8 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/Helper/Payolution.php(130): Payone_Core_Helper_Payolution->isPayolutionMethodAvailable()
#9 [internal function]: Payone_Core_Helper_Payolution->getPayolutionFraudPreventionJs('')
#10 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Model/Layout.php(324): call_user_func_array(Array, Array)
#11 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Model/Layout.php(214): Mage_Core_Model_Layout->_generateAction(Object(Mage_Core_Model_Layout_Element), Object(Mage_Core_Model_Layout_Element))
#12 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Model/Layout.php(210): Mage_Core_Model_Layout->generateBlocks(Object(Mage_Core_Model_Layout_Element))
#13 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Controller/Varien/Action.php(344): Mage_Core_Model_Layout->generateBlocks()
#14 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Controller/Varien/Action.php(269): Mage_Core_Controller_Varien_Action->generateLayoutBlocks()
#15 /var/www/shops/magento/1942ds/app/code/core/Mage/Checkout/controllers/OnepageController.php(215): Mage_Core_Controller_Varien_Action->loadLayout()
#16 /var/www/shops/magento/1942ds/app/code/community/Payone/Core/controllers/Checkout/OnepageController.php(366): Mage_Checkout_OnepageController->indexAction()
#17 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Controller/Varien/Action.php(418): Payone_Core_Checkout_OnepageController->indexAction()
#18 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Controller/Varien/Router/Standard.php(254): Mage_Core_Controller_Varien_Action->dispatch('index')
#19 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Controller/Varien/Front.php(172): Mage_Core_Controller_Varien_Router_Standard->match(Object(Mage_Core_Controller_Request_Http))
#20 /var/www/shops/magento/1942ds/app/code/core/Mage/Core/Model/App.php(365): Mage_Core_Controller_Varien_Front->dispatch()
#21 /var/www/shops/magento/1942ds/app/Mage.php(690): Mage_Core_Model_App->run(Array)
#22 /var/www/shops/magento/1942ds/index.php(83): Mage::run('', 'store')
#23 {main}";s:3:"url";s:39:"/shops/magento/1942ds/checkout/onepage/";s:11:"script_name";s:31:"/shops/magento/1942ds/index.php";s:4:"skin";s:7:"default";}
```

sprankhub commented:
Shit. Any idea regarding this issue @mwr? :)

mwr commented:
We are evaluating the upgrade to 4.4.0 this week and next week. 
I am planning on updating my pull request to 4.4.0 and using that 4.4.0+pull-349 as the version we are upgrading to.
I'll let you know how that went.

mwr commented:
@fjbender @sprankhub I have just pushed an update to this pull-request, which addresses the "registry key already exists" exception. So far I was able to create orders. Just now I am about to deploy it to our staging environment and continue testing there.

mwr commented:
It seems like the fix worked. But I cannot say for sure, since we are having other issues with Creditcard payments after the upgrade to the latest release.
Therefore we cannot check all payment methods and the complete payment process.

T-Kuchel commented:
Thx @mwr !

mwr commented:
Today, we rolled out the PAYONE Module in version 4.4.0 including this PR. 
The last 2 weeks the upgrade incl. the fix was tested and no errors were found.
Will get back to this PR if anything changes
